### PR TITLE
use time step embedding from file

### PR DESCRIPTION
**Pull request by RSMNYS**
Timestamp-embedding-parser

### CODE REVIEW

**Comment by freedomtan** [comment]
@mohitmundhragithub to share the code used to generate this embedding.

**Comment by mohitmundhragithub** [comment]
@RSMNYS, The function where the time step embedding is generated is shared here: https://github.com/mlcommons/submissions_mobile_v4.1/issues/6#issuecomment-2399205941

> From QPM Tutorial package, if we go to this file: <qpm_package>\StableDiffusion1.5\model\example1\redefined_modules\diffusers\models\embeddings.py
get_timestep_embedding( ) function generates the time embeddings.

> <qpm_package>\StableDiffusion1.5\model\example3\host_linux_target_android_without_native\qnn_model_execution.ipynb
Above function is called from the top level notebook: get_time_embedding( )

image

**Comment by RSMNYS** [comment]
> @RSMNYS, The function where the time step embedding is generated is shared here: https://github.com/mlcommons/submissions_mobile_v4.1/issues/6#issuecomment-2399205941
> 
> > From QPM Tutorial package, if we go to this file: <qpm_package>\StableDiffusion1.5\model\example1\redefined_modules\diffusers\models\embeddings.py
> > get_timestep_embedding( ) function generates the time embeddings.
> 
> > <qpm_package>\StableDiffusion1.5\model\example3\host_linux_target_android_without_native\qnn_model_execution.ipynb
> > Above function is called from the top level notebook: get_time_embedding( )
> 
> image

@mohitmundhragithub Shown code is only for the embedding generation. And no operations from the cut part are shown. (fully connected layer, Logistic/Sigmoid activation, and mul operation.

**Comment by freedomtan** [comment]
@AhmedTElthakeb to provide a small tflite with only 3 operations.

@mohitmundhragithub will try to compare tflite and onnx files.

**Comment by AhmedTElthakeb** [comment]
> @AhmedTElthakeb to provide a small tflite with only 3 operations.
[sdv15_time_emb_head_int8.zip](https://github.com/user-attachments/files/17963393/sdv15_time_emb_head_int8.zip)

**Comment by RSMNYS** [comment]
Hi guys! So I've used to the model prepared by Ahmed to generate the embedding file. Now all works good. Here the link to the colab: https://colab.research.google.com/drive/1LG_rC5dlx2CbW2ZF4EamOnCdwtwLQESv?usp=sharing to prepare the embedding file, and the link to the models and embedding file: https://drive.google.com/drive/folders/1CT6VUWwGaTw34Za6dTJE7ptms20NUKH3?usp=sharing

**Comment by freedomtan** [comment]
please use .pkl for the save the embedding.

@mohitmundhragithub please help figure out the difference between the one generated by @RSMNYS and Q's.

**Comment by freedomtan** [comment]
once the .pkl is done, please make the embedding part of the files to be downloaded for Stable Diffusion for the TFLite backend.

**Comment by RSMNYS** [comment]
> please use .pkl for the save the embedding.
> 
> @mohitmundhragithub please help figure out the difference between the one generated by @RSMNYS and Q's.

here you can find 2 files in json format to be able to compare numbers at least: https://drive.google.com/drive/folders/1SO1akyvWd2uYz9Xf_u5OGBJLetuW6x5A?usp=sharing

**Comment by RSMNYS** [comment]
guys, I've updated the colab to generate the pkl file. Also adjusted embedding_utils code to parse this. Here you can find the pkl file: https://drive.google.com/file/d/1pDd5wZje1KbIS4JcWzhDx00aN8GmTBpc/view?usp=share_link

**Comment by RSMNYS** [comment]
> once the pickle is done, please make the embedding part of the files to be downloaded for Stable Diffusion for the TFLite backend.

@anhappdev I think for this we only need to upload the new UNET model (without embedding operations), and our embedding file. Right?  If yes, can you please help upload those 2 files to the storage, from where we download models and other assets:

https://drive.google.com/file/d/1pDd5wZje1KbIS4JcWzhDx00aN8GmTBpc/view?usp=share_link
https://drive.google.com/file/d/1Sf2lcRDjSfg9jgABWWbV5EeWEmXl5CsJ/view?usp=share_link

**Comment by anhappdev** [comment]
> @anhappdev I think for this we only need to upload the new UNET model (without embedding operations), and our embedding file. Right? If yes, can you please help upload those 2 files to the storage, from where we download models and other assets:
>
>https://drive.google.com/file/d/1pDd5wZje1KbIS4JcWzhDx00aN8GmTBpc/view?usp=share_link
>https://drive.google.com/file/d/1Sf2lcRDjSfg9jgABWWbV5EeWEmXl5CsJ/view?usp=share_link

@RSMNYS Here is the URL for the 2 files you shared 
(Please remember to update the checksum in the backend settings)
https://mobile.mlcommons-storage.org/app-resources/models/v4_1/tflite/timestep_embeddings_data.pkl
https://mobile.mlcommons-storage.org/app-resources/models/v4_1/tflite/sd_diffusion_model_dynamic.tflite


And other model files for reference:
https://mobile.mlcommons-storage.org/app-resources/models/v4_1/tflite/sd_decoder_dynamic.tflite
https://mobile.mlcommons-storage.org/app-resources/models/v4_1/tflite/sd_text_encoder_dynamic.tflite

**Comment by Mostelk** [comment]
We need to update the release here with the new model and embeddings https://github.com/mlcommons/mobile_open/releases

**Comment by anhappdev** [comment]
> We need to update the release here with the new model and embeddings https://github.com/mlcommons/mobile_open/releases

- [ ] Ok. I will do it after we merge this PR.

File: .fvm/flutter_sdk
Comment (anhappdev) [inline .fvm/flutter_sdk:1]: Please add the `.fvm` directory to `.gitignore`.

**Comment by mohitmundhragithub** [comment]
> 2 files to the storage, fr

I compared the 2 TS embeddings. 
Few observations:

1. The timestamp sequences in the 2 cases are different. QTI backend has the timestep sequences from here: https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_dpmsolver_multistep.py#L364
    TFLite uses the the TS values directly from the tflite model. QTI uses it from the DPMSolverMultistepScheduler.  The end result being values of the time stamps are different in the 2 cases. 
image
2. There are some operations done on the above TimeStamp values, in the function get_time_embedding() in the file model\example1\redefined_modules\diffusers\models\unet_2d_condition.py, which in turn calls the function shared here: https://github.com/mlcommons/mobile_app_open/pull/928#issuecomment-2482672443 . These operations are required for the QTI backend's UNet's latent input.   
3. Even on using the same TS values as got from DPMSolverMultistepScheduler, and applying them using the logic shared by Sergji in https://drive.google.com/file/d/1pDd5wZje1KbIS4JcWzhDx00aN8GmTBpc/view?usp=share_link, to generate the embeddings, QTI backend doesn't get acceptable accuracy.
4. I would like to reiterate that the Latent inputs in this case are dependent on the platform / sw stack / quantization methods used by each backend etc. We can't have the same latent inputs for each of the platform, and hence we shouldn't make it part of the app. It should be part of the backend instead. For these large models we need to accept this fact that the implementation for each of the backend can't be exactly the same. Right now Stable Diffusion and going fwd other LLMs is going to have these limitations. Moreover, the latent input is not the input to the use case, but it's the input of one of the intermediate models.
5. I don't see the TS embeddings being separately defined in the inference group as well.
6. The Unet model that is loaded by the Tutorial script is shared here at this location inside the QPM: model\example1\redefined_modules\diffusers\models\unet_2d_condition.py. Its creating the same model (minus the three ops) using the original weights but also, adopting to the AIMET quantization for quantsim. This model is then exported to the onnx format, which is consumed during inference.

**Comment by mohitmundhragithub** [comment]
> Why is QTI backend using a different scheduler than one in the reference implementation? We have commit id in the app.
We are happy to align to that scheduler or QTI should use same scheduler, let us discuss the differences in the meeting

QTI backend is using the standard scheduler, as has been prescribed in the original model, and also the scheduler implementation in huggingface as already mentioned earlier (standard torch implementation).
https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_dpmsolver_multistep.py#L364

This is called from stable_diffusion.ipynb (line 153) in the tutorial from QPM.
`from diffusers import DPMSolverMultistepScheduler\n`

> Please provide a PR to show the exact extra operations that are missing in the TimeStamp calculation, are these operations done alternatively in the network for reference implementation?

image

This is the exact function used to find the TimeStepEmedding. I see the exact same functions also defined in here in the huggingface repository for diffusers: 
https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/embeddings.py#L976
Basically, its linear function -> silU -> linear function again.

> Did you quantize again, or just pass it through?

Didn't quantize again. Just passthrough.

Seems like there are a few differences in the original pytorch model from huggingface and the tflite model?

> We are happy to align with QTI, can you do a PR that enable us to reproduce the embedding? This is an input to the network, not an intermediate latent, so everyone should use same input to have a Fair benchmark

The input to the whole SD pipeline is the input prompt. Because of the complexity of the implementation, different frameworks seems to have different implementation in this case.

**Comment by freedomtan** [comment]
@mohitmundhragithub please confirm what the input is supposed to be.
@AhmedTElthakeb: suggested the op / layer after the timestep embeddeding is SiLU, which mean, it cut after the second linear layer.

@freedomtan to check the two schedulers (the one used by Keras, the one used by Diffuser).
@RSMNYS: please the two schedulers, too.

**Comment by anhappdev** [comment]
Should we merge this PR into `submission-v4.1` instead of `master`?

**Comment by mohitmundhragithub** [comment]
> @mohitmundhragithub please confirm what the input is supposed to be. @AhmedTElthakeb: suggested the op / layer after the timestep embeddeding is SiLU, which mean, it cut after the second linear layer.
> 
> @freedomtan to check the two schedulers (the one used by Keras, the one used by Diffuser). @RSMNYS: please the two schedulers, too.

I checked the onnx file again. input_2 is fed to sigmoid and mul which should be equivalent to SiLU as discussed in the meeting today.

**Comment by mohitmundhragithub** [comment]
> Should we merge this PR into `submission-v4.1` instead of `master`?

I think so. 
Master doesn't have changes from qti backend from submission-v4.1 as well, yet.

**Comment by Mostelk** [comment]
@RSMNYS Please check the new models shared by @AhmedTElthakeb in mobile_model_closed to generate the new time emebeddings

**Comment by AhmedTElthakeb** [comment]
@RSMNYS 
Release: https://github.com/mlcommons/mobile_model_closed/releases/tag/alpha-tflite-v0.2 

1. Time embedding head:` Lin -> SiLU -> Lin` [float32 model] 
`sdv15_time_emb_head_LinSiluLin_fp32.tflite`
2. UNET new time embedding input goes to `SiLU` layer [int8 model]
`sd_diffusion_full_model_dynamic_int8_mlperfv41_v2.tflite`

Also here: https://github.com/mlcommons/mobile_model_closed/issues/12#issuecomment-2533334894

**Comment by freedomtan** [comment]
Current noise scheduling for TFLite was derived from my C++ code, which in turn was from Keras CV's code
1. according to the Keras CV Stable Diffusion [noise scheduler code and comments](https://github.com/keras-team/keras-cv/blob/master/keras_cv/src/models/stable_diffusion/noise_scheduler.py), it was derived from Hugging Face DDPM implementation
2. it seems to me what @mohitmundhragithub referred to is not  a scheduling method, but a ODE solver.

**Comment by RSMNYS** [comment]
> @RSMNYS Release: https://github.com/mlcommons/mobile_model_closed/releases/tag/alpha-tflite-v0.2
> 
> 1. Time embedding head:` Lin -> SiLU -> Lin` [float32 model]
>    `sdv15_time_emb_head_LinSiluLin_fp32.tflite`
> 2. UNET new time embedding input goes to `SiLU` layer [int8 model]
>    `sd_diffusion_full_model_dynamic_int8_mlperfv41_v2.tflite`
> 
> Also here: [mlcommons/mobile_model_closed#12 (comment)](https://github.com/mlcommons/mobile_model_closed/issues/12#issuecomment-2533334894)

Hi guys! The new models works good. But we have one issue with pickle file format. And the problem is  it's hard to decode and parse in c++, we need to use some special libs to parse it (and I wan't able to find such), or manual parsing which is hard and complex to parse the pickle data and protocols. So what I did I saved the tilmestep and embeddings in the same format as we originally had from Mohit. So while we can have the embedding part in pkl file format, in the c++ code I still use the old bin file format. 

So let's decide tomorrow if we need to stick to pkl or we can use the existing one .bin. 

The code to save embedding you can see in colab:

https://colab.research.google.com/drive/1LG_rC5dlx2CbW2ZF4EamOnCdwtwLQESv#scrollTo=B6UTUZcQuj2f

**Comment by mohitmundhragithub** [comment]
we also tried at our end. 
The output generated with the timestep sequences from AIMET and the embeddings generated using the script which Sergji shared, works fine and gives good accuracy.
we still want to try generating the accruacy score with the timestep sequences from tflite.

**Comment by mohitmundhragithub** [comment]
Hi, 

This is the script that we used to generate the pickle file. It contains couple of changes to generate the .pkl file in required format. 
[time_emeddings.py.txt](https://github.com/user-attachments/files/18164356/time_emeddings.py.txt)

I was slightly wrong in the meeting. The qti backend cpp code doesn't consume the pkl file directly. The pkl file is consumed by the flatten.py file shared here: https://github.com/mlcommons/mobile_app_open/blob/submission-v4.1/mobile_back_qti/DLC/util/StableDiffusion/flatten.py#L96

This generates a bin file, which gets packed along with other model files, and is consumed by the qti backend along with the models.
@RSMNYS was also suggesting something similar in the meeting.

**Comment by freedomtan** [comment]
@anhappdev please help put the embedding file to the right place (Cloudflare?)
@mohitmundhragithub please share the "final" embedding binary file (so that we can double-check it)

@anhappdev please create 5.0 submission branch after the checking is done. And then we can fixed the file location, and then free the submission branch.

**Comment by anhappdev** [comment]
@RSMNYS Please change the base branch of this PR from `master` to `submission-v4.1`.

**Comment by anhappdev** [comment]
@RSMNYS Can you please create a public GitHub repo and upload all the relevant files there like I did in [this one](https://github.com/anhappdev/tmp/releases) and use those GitHub links for the backend settings.

After we have a running code with final models / files, I will replace the GitHub links with the CloudFare one. This way I don't miss or upload wrong files.

**Comment by mohitmundhragithub** [comment]
> @mohitmundhragithub please share the "final" embedding binary file (so that we can double-check it)

[unet_time_step_embeddings_20.pkl.txt](https://github.com/user-attachments/files/18292681/unet_time_step_embeddings_20.pkl.txt)
[timestep_steps_20_int32_embedding_1x1280_float32.bin.ts.txt](https://github.com/user-attachments/files/18292686/timestep_steps_20_int32_embedding_1x1280_float32.bin.ts.txt)

Sharing the .pkl files and the .bin files generated as described in the steps here: 
https://github.com/mlcommons/mobile_app_open/pull/928#issuecomment-2548150726

Please note that i was unable to upload .pkl and .bin.ts files (github's restriction), so had to append .txt in the filenames. Please remove the .txt extension to use those.

**Comment by RSMNYS** [comment]
> > @mohitmundhragithub please share the "final" embedding binary file (so that we can double-check it)
> 
> [unet_time_step_embeddings_20.pkl.txt](https://github.com/user-attachments/files/18292681/unet_time_step_embeddings_20.pkl.txt) [timestep_steps_20_int32_embedding_1x1280_float32.bin.ts.txt](https://github.com/user-attachments/files/18292686/timestep_steps_20_int32_embedding_1x1280_float32.bin.ts.txt)
> 
> Sharing the .pkl files and the .bin files generated as described in the steps here: [#928 (comment)](https://github.com/mlcommons/mobile_app_open/pull/928#issuecomment-2548150726)
> 
> Please note that i was unable to upload .pkl and .bin.ts files (github's restriction), so had to append .txt in the filenames. Please remove the .txt extension to use those.

that works good!

**Comment by RSMNYS** [comment]
Hi guys! I can run stable diffusion on my android device (Samsung Galaxy S22), but at the end I have 0 as a result, however I see all the steps in logs. Checking why. Also, during the stable diffusion process we are not updating the progress, so it's hard to understand what is going on.

**Comment by mohitmundhragithub** [comment]
> Hi guys! I can run stable diffusion on my android device (Samsung Galaxy S22), but at the end I have 0 as a result, however I see all the steps in logs. Checking why. Also, during the stable diffusion process we are not updating the progress, so it's hard to understand what is going on.

can you share the loadgen and logcat logs?

**Comment by freedomtan** [comment]
Let's 
1. make sure this app works well (check the app log directory, @RSMNYS )
2. verify tflite files are the same with what @AhmedTElthakeb provided (https://github.com/mlcommons/mobile_model_closed/releases/tag/alpha-tflite-v0.3)
3. after we merge this, add tflite files, .ts, files to both the mobile_models file and the CloudFlare bucket (@anhappdev)
4. add .pickle and tflite files to the mobile_open repo (https://github.com/mlcommons/mobile_open/releases). @anhappdev

**Comment by RSMNYS** [comment]
[Benchmark result.json](https://github.com/user-attachments/files/18329721/Benchmark.result.json)
here is my benchmark result file

**Comment by anhappdev** [comment]
I tested this PR. Here is the output images of the TFLite backend:
https://drive.google.com/drive/folders/1-2OBHMlfF3fbuXcejeIQzPE-2mD5KPsU?usp=sharing

Many images look "wrong" with the color like this one, please check them out.


Noted that the output images are only saved to disk when we run the accuracy test. 
The images should be in `
/storage/emulated/0/Android/data/org.mlcommons.android.mlperfbench/files/logs/<timestamp>/stable_diffusion-accuracy/cocogen_outputs/
`

**Comment by mohitmundhragithub** [comment]
> I tested this PR. Here is the output images of the TFLite backend: https://drive.google.com/drive/folders/1-2OBHMlfF3fbuXcejeIQzPE-2mD5KPsU?usp=sharing
> 
> Many images look "wrong" with the color like this one, please check them out. 
> 
> Noted that the output images are only saved to disk when we run the accuracy test. The images should be in `/storage/emulated/0/Android/data/org.mlcommons.android.mlperfbench/files/logs/<timestamp>/stable_diffusion-accuracy/cocogen_outputs/`

almost all the images have such artifacts. 
seems more like the output is clipped to some max / min values for some pixels.

**Comment by freedomtan** [comment]
> I tested this PR. Here is the output images of the TFLite backend: https://drive.google.com/drive/folders/1-2OBHMlfF3fbuXcejeIQzPE-2mD5KPsU?usp=sharing
> 
> Many images look "wrong" with the color like this one, please check them out. 
> 
> Noted that the output images are only saved to disk when we run the accuracy test. The images should be in `/storage/emulated/0/Android/data/org.mlcommons.android.mlperfbench/files/logs/<timestamp>/stable_diffusion-accuracy/cocogen_outputs/`

Most likely, it's caused by quantization. At least, we know timesteps embedding works as expected. 
To further confirm which model (text encoder, unet/diffusion, or image decoder) is problematic, we can check fp32/fp16 text encoder and image decoder first.

**Comment by freedomtan** [comment]
Let's merge this and create a 5.0 submission branch, then work on using fp16 models.

@AhmedTElthakeb please check if you can export fp16 tflite models (https://ai.google.dev/edge/litert/models/post_training_float16_quant).

**Review by freedomtan — APPROVED** [review]

**Comment by freedomtan** [comment]
By 5.1, qualcomm backend will use the timestep binary file (so that we have a shared timestep source file). @mohitmundhragithub

**Comment by AhmedTElthakeb** [comment]
Here is the fp16 versions release: [alpha-tflite-v0.4](https://github.com/mlcommons/mobile_model_closed/releases/tag/alpha-tflite-v0.4)
